### PR TITLE
Adds a secret checksum annotation to deployment

### PR DESCRIPTION
**Pull request by DhritiShikhar**
Issue Link -> [CFE-366](https://issues.redhat.com/browse/CFE-366)

Changes:

1. Adds a watch for target secret
2. Creates hash of target secret data
3. Adds hash to deployment's annotation
Example: `externaldns.olm.openshift.io/credentials-secret-hash: 72ddc0c540ab3a290e138726940591debf14c581`

### CODE REVIEW

File: pkg/operator/controller/externaldns/deployment.go [hunk 1]
```diff
@@ -265,8 +319,21 @@ func (r *reconciler) updateExternalDNSDeployment(ctx context.Context, current, d
...
+// externalDNSSecretChanged returns true if the current secret annotation differ from the expected
+func externalDNSSecretChanged(current, expected, updated *appsv1.Deployment) bool {
+	changed := false
+	for expected_key, expected_val := range expected.Annotations {
+		if current_val, ok := current.Annotations[expected_key]; ok {
+			if current_val != expected_val {
+				updated.Annotations[expected_key] = expected_val
+				return true
+			}
+		}
```
Comment (sherine-k) [inline pkg/operator/controller/externaldns/deployment.go:334]: Hi @DhritiShikhar 
If the annotation is not found, we should consider the secret as changed as well, right?

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 2]
```diff
@@ -58,6 +58,7 @@ var (
 )
 
 func TestDesiredExternalDNSDeployment(t *testing.T) {
+	secretHash := "f17bceb5a060e33473c68229903ef5c517d9a172"
```
Comment (sherine-k) [inline pkg/operator/controller/externaldns/deployment_test.go:61]: nit: a suggestion would be to make this a constant, as it is used in several places.

**Review by sherine-k — COMMENTED** [review]
Thanks @DhritiShikhar 
This PR looks good except from some nitpicks.
I have a doubt what should be done when the secret is not found: do we deploy with an empty hash?

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 3]
```diff
@@ -2259,6 +2261,17 @@ func TestExternalDNSDeploymentChanged(t *testing.T) {
...
+			description: "if externalDNS annotation changes",
+			originalDeployment: testDeploymentWithContainers([]corev1.Container{
+				testContainer(),
+			}),
+			mutate: func(dep1 *appsv1.Deployment) {
+				dep1.Annotations = updatedSecretHashAnnotation
+			},
+			expect:             true,
+			expectedDeployment: testDeploymentWithAnnotations(updatedSecretHashAnnotation),
+		},
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:2274]: Can you please add a test case when the secret annotation didn't exist in the current deployment?

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 4]
```diff
@@ -3161,6 +3181,12 @@ func testDeploymentWithContainers(containers []corev1.Container) *appsv1.Deploym
 	return depl
 }
 
+func testDeploymentWithAnnotations(updatedSecretHashAnnotation map[string]string) *appsv1.Deployment {
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:3184]:
```suggestion
func testDeploymentWithAnnotations(annotations map[string]string) *appsv1.Deployment {
```

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 5]
```diff
@@ -2771,6 +2787,7 @@ func TestEnsureExternalDNSDeployment(t *testing.T) {
 							BlockOwnerDeletion: &test.TrueVar,
 						},
 					},
+					Annotations: map[string]string{"external-dns-credentials-test": ""},
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:2790]: Can you please add a test case with an actual secret in `existingObjects` so that the hash calculation is tested?

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 6]
```diff
@@ -3125,10 +3142,13 @@ func TestEnsureExternalDNSDeployment(t *testing.T) {
 }
 
 func testDeployment() *appsv1.Deployment {
+	secretHashAnnotation := make(map[string]string)
+	secretHashAnnotation["external-dns-credentials-test"] = "f17bceb5a060e33473c68229903ef5c517d9a172"
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:3146]: This annotation comes a bit out of nowhere. There is dedicated function for the test deployment with annotation, this function can be used when an annotation is needed.

File: pkg/operator/controller/externaldns/deployment.go [hunk 7]
```diff
@@ -114,6 +130,39 @@ func (r *reconciler) ensureExternalDNSDeployment(ctx context.Context, namespace,
 	return true, current, nil
 }
 
+// buildSecretHash is a utility function to get a checksum of the resource data
+func buildSecretHash(data map[string][]byte) (string, error) {
```
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go]: Done here cf8b1ab9a4c96f34ff2482f4ea643758d0089027

File: pkg/operator/controller/externaldns/deployment.go [hunk 8]
```diff
@@ -114,6 +130,39 @@ func (r *reconciler) ensureExternalDNSDeployment(ctx context.Context, namespace,
...
+		_, err = hash.Write(data[k])
+		if err != nil {
+			return "", err
+		}
+	}
+	return hex.EncodeToString(hash.Sum(nil)), nil
+}
+
+// currentExternalDNSSecret gets the current externalDNS secret resource.
+func (r *reconciler) currentExternalDNSSecret(ctx context.Context, nsName types.NamespacedName, secretName string) (bool, *corev1.Secret, error) {
```
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go]: Removed here cf8b1ab9a4c96f34ff2482f4ea643758d0089027

File: pkg/operator/controller/externaldns/deployment.go (hunk 1 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go]: Resolved here 578de6b1904c7a4cf790b72ef7f201ae6b04de65

File: pkg/operator/controller/externaldns/deployment.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go]: Resolved

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 2 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go]: Resolved here 4725c696ebfa116c8147cc3bd98362e8bc76c64f

File: pkg/operator/controller/externaldns/deployment.go [hunk 9]
```diff
@@ -265,8 +319,21 @@ func (r *reconciler) updateExternalDNSDeployment(ctx context.Context, current, d
 // Returns a boolean if an update is necessary, and the deployment resource to update to.
 func externalDNSDeploymentChanged(current, expected *appsv1.Deployment) (bool, *appsv1.Deployment) {
 	updated := current.DeepCopy()
+	return externalDNSSecretChanged(current, expected, updated) || externalDNSContainersChanged(current, expected, updated), updated
+}
 
-	return externalDNSContainersChanged(current, expected, updated), updated
+// externalDNSSecretChanged returns true if the current secret annotation differ from the expected
+func externalDNSSecretChanged(current, expected, updated *appsv1.Deployment) bool {
```
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go]: Resolved 984cfd7d76eacee68595608aa534f070af997a5e

File: pkg/operator/controller/externaldns/deployment.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go:312]: Resolved 984cfd7d76eacee68595608aa534f070af997a5e

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 5 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go]: Resolved

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 6 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go]: Resolved 7734cdd74f2384679279d20f0130bcc9d728d821

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 3 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go:2275]: test Added

File: pkg/operator/controller/externaldns/deployment.go [hunk 10]
```diff
@@ -265,8 +296,24 @@ func (r *reconciler) updateExternalDNSDeployment(ctx context.Context, current, d
...
+		return true
+	}
+	for expectedKey, expectedValue := range expected.Annotations {
+		currentVal, currentExists := current.Annotations[expectedKey]
+		if !currentExists || currentVal != expectedValue {
+			updated.Annotations[expectedKey] = expectedValue
+			return true
+		}
+	}
+	return changed
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment.go:316]:
```suggestion
	return false
```
`changed` variable is used only here, can return `false` directly.

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 11]
```diff
@@ -2290,6 +2313,7 @@ func TestEnsureExternalDNSDeployment(t *testing.T) {
 		name               string
 		existingObjects    []runtime.Object
 		expectedExist      bool
+		ExpectedSecret     corev1.Secret
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:2316]: This field is not used in the test case, should be removed.

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 12]
```diff
@@ -2363,11 +2394,13 @@ func TestEnsureExternalDNSDeployment(t *testing.T) {
 					},
 				},
 			},
+			errExpected: true,
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:2397]: Why `Doesn't exist` test case became an error case?
If the deployment doesn't exist it has to be created and match the desired deployment. You just need to add `testSecret()` into existing objects as for the other test cases.

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 13]
```diff
@@ -3029,6 +3076,7 @@ func TestEnsureExternalDNSDeployment(t *testing.T) {
 							BlockOwnerDeletion: &test.TrueVar,
 						},
 					},
+					Annotations: map[string]string{credentialsAnnotation: testSecretHash},
 				},
 				Spec: appsv1.DeploymentSpec{
 					Replicas: &replicas,
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:3082]: Can you please add a test case when no secret existed and `ensureDeployment` failed due to this?

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 14]
```diff
@@ -3377,3 +3440,45 @@ func testPlatformStatusGCP(projectID string) *configv1.PlatformStatus {
 		},
 	}
 }
+
+func TestBuildSecretHash(t *testing.T) {
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment_test.go:3444]: Can you please move this golang test function to the other golang test functions?
Just to keep it a bit organized because now we have: `TestX`, `TestY`, `testDeployment`, `testSecret`, `TestZ`.

File: pkg/operator/controller/externaldns/status_test.go [hunk 15]
```diff
@@ -557,7 +557,7 @@ func fakeExternalDNS() *operatorv1alpha1.ExternalDNS {
 				Type: operatorv1alpha1.ProviderTypeAWS,
 				AWS: &operatorv1alpha1.ExternalDNSAWSProviderOptions{
 					Credentials: operatorv1alpha1.SecretReference{
-						Name: testSecretName,
+						Name: test.OperandName,
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/status_test.go:560]: Why this change? And why it didn't impact any before existing tests?

**Review by alebedev87 — COMMENTED** [review]
Thanks for addressing the previous comments. Some final remarks.

File: pkg/operator/controller/externaldns/deployment.go [hunk 16]
```diff
@@ -72,15 +75,26 @@ var sourceStringTable = map[operatorv1alpha1.ExternalDNSSourceType]string{
...
 	nsName := types.NamespacedName{Namespace: namespace, Name: controller.ExternalDNSResourceName(externalDNS)}
-
-	secretName := controller.ExternalDNSDestCredentialsSecretName(r.config.Namespace, externalDNS.Name).Name
 	configMapName := ""
 	if r.config.InjectTrustedCA {
 		configMapName = controller.ExternalDNSDestTrustedCAConfigMapName("").Name
 	}
-	desired, err := desiredExternalDNSDeployment(namespace, image, secretName, serviceAccount, externalDNS, r.config.IsOpenShift, r.config.PlatformStatus, configMapName)
+
+	secretExists, secret, err := r.currentExternalDNSSecret(ctx, nsName)
```
Comment (alebedev87) [inline pkg/operator/controller/externaldns/deployment.go:84]: This will not work, deployment name != secret name.
```suggestion
	secretExists, secret, err := r.currentExternalDNSSecret(ctx, controller.ExternalDNSDestCredentialsSecretName(namespace, externalDNS.Name))
```
I think that e2e tests failed because of this: no secret was found - no externaldns operand was created.

File: pkg/operator/controller/externaldns/status_test.go (hunk 15 above)
Comment (alebedev87) [reply pkg/operator/controller/externaldns/status_test.go:560]: Sorry I missed the change in `pkg/operator/controller/externaldns/controller_test.go` where the secren name was replaced with the operand name. That answers the second question. But I'm still curious why the secret name was replaced by the operand. Maybe because of this https://github.com/openshift/external-dns-operator/pull/127#discussion_r848920447.

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 11 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go]: Done here 6f4d092af25d1c2fc93cb34347d5e03f794cf83d

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 12 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go]: Done here 6f4d092af25d1c2fc93cb34347d5e03f794cf83d

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 14 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go:3443]: @alebedev87 Where exactly do you suggest that I move this test function?

File: pkg/operator/controller/externaldns/status_test.go (hunk 15 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/status_test.go]: Removed the change

**Review by sherine-k — APPROVED** [review]
Nice work @DhritiShikhar

File: pkg/operator/controller/externaldns/deployment.go [hunk 17]
```diff
@@ -265,8 +295,23 @@ func (r *reconciler) updateExternalDNSDeployment(ctx context.Context, current, d
...
+func externalDNSAnnotationsChanged(current, expected, updated *appsv1.Deployment) bool {
+	if current.Annotations == nil {
+		updated.Annotations = expected.Annotations
+		return true
+	}
+	for expectedKey, expectedValue := range expected.Annotations {
+		currentVal, currentExists := current.Annotations[expectedKey]
+		if !currentExists || currentVal != expectedValue {
+			updated.Annotations[expectedKey] = expectedValue
+			return true
```
[the remark this answers is not in the document]
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go]: Updated this @sherine-k

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 14 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go]: Moved it here https://github.com/openshift/external-dns-operator/pull/127/files#diff-fb0a8674418558ba280aa06686d117445fa00bf83fdca12efc520d8836460089R3174

Comment (alebedev87) [reply pkg/operator/controller/externaldns/deployment_test.go]: Just where I though, thanks a lot, @DhritiShikhar!

File: pkg/operator/controller/externaldns/controller_test.go [hunk 18]
```diff
@@ -289,7 +289,7 @@ func testExtDNSInstance() *operatorv1alpha1.ExternalDNS {
 func testSecret() *corev1.Secret {
 	return &corev1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      testSecretName,
+			Name:      "external-dns-credentials-test",
```
Comment (lmzuccarelli) [inline pkg/operator/controller/externaldns/controller_test.go:292]: nit: maybe declare a constant for all literals that are duplicated in the code i.e "deployment"  , "external-dns-test" (3 times) , "serviceaccount" (6 times),  "externaldns" (3 times). It helps when refactoring code (i.e change in one place only) and neatens code

File: pkg/operator/controller/externaldns/deployment.go [hunk 19]
```diff
@@ -127,7 +154,7 @@ func (r *reconciler) currentExternalDNSDeployment(ctx context.Context, nsName ty
 }
 
 // desiredExternalDNSDeployment returns the desired deployment resource.
-func desiredExternalDNSDeployment(namespace, image, secretName string,
+func desiredExternalDNSDeployment(namespace, image, secretName string, secretHash string,
```
Comment (lmzuccarelli) [inline pkg/operator/controller/externaldns/deployment.go:157]: wow 9 parameters - should you consider using a new struct ? its not a big deal, usually in go with more than 4 params you should look at creating a new struct

File: pkg/operator/controller/externaldns/deployment_test.go [hunk 20]
```diff
@@ -57,6 +57,8 @@ var (
 	}
 )
 
+const testSecretHash = "e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855"
```
Comment (lmzuccarelli) [inline pkg/operator/controller/externaldns/deployment_test.go:60]: this is good - can you create constants  for all duplicated literals in this file

**Review by lmzuccarelli — COMMENTED** [review]
Overall looks good - could you address the nit issues ?

File: pkg/operator/controller/externaldns/controller_test.go (hunk 18 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/controller_test.go]: Done here 8d680dc2c2f2b12800f60a5326c97e283d80fadc

File: pkg/operator/controller/externaldns/deployment.go (hunk 19 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment.go]: Done here ac35fea8394353bf3ae40e85fc716e258b203025

File: pkg/operator/controller/externaldns/deployment_test.go (hunk 20 above)
Comment (DhritiShikhar) [reply pkg/operator/controller/externaldns/deployment_test.go]: Done here v9dd747d85d0c90a7e7c4b1f29746f26be9ee9f2d
